### PR TITLE
[SR GeekCamp]  Support cluster management in new Parser and new Analyzer.

### LINKED ISSUE #4044: New Parser and Analyzer support cluster management statements【Easy】
# Requirement Description
The new Parser and the new Analyzer support cluster management statements

# Mentor
@HangyuanLiu 

# Registration procedure

**Step 1.** Add a comment when you sign up. The comment can be a brief self-introduction and reason for signing up, including what you do, what you're good at, what you're interested in and your technology stack, and how familiar you are with StarRocks. Example:

> I pick! My name is Johnny. I am from StarRocks and I have been working on database kernel development for 1 year. I am interested in OLAP, good at C++ programming, and have been using StarRocks for 6 months. I chose to solve this issue because I am familiar with execution engine.

**Step 2.** On your WeChat, add the StarRocks community assistant "**_StarRocks-1_**" as a WeChat friend and leave your GitHub ID. The assistant will invite you to a WeChat group based on the issue you picked.

**Step 3.** The mentor will communicate with all the participants who picked this issue and assign the task to the best match. 

# Before you start
1. Read the StarRocks [GitHub workflow](https://github.com/StarRocks/community/blob/main/Contributors/guide/workflow.md) carefully, which provides the configurations of BE and FE and describes how to submit a pull request.
2. To learn more about StarRocks, read [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).

# Pull request requirements

1. After you complete code development, upload the code to the StarRocks code repo as a pull request (PR).

  - PR title format: **[SR GeekCamp] + PR title**
  - Labels ：[Geek Camp 2022/feature]
2. After you submit your PR, comment on the status of your PR submission as requested under the Pinned Issue [[StarRocks geek Camp 2022] ](https://github.com/StarRocks/starrocks/issues/4008)page.

  - Comment format: **"I have finished Issue #" + PR link id**

### PR DESCRIPTION
**Pull request by happut**
- Add/DropBackend
- Add/DropObserver
- Add/DropFollower

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4044

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 26 / 48 (54.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/FrontendClause.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterSystemStmt.java|1|5|20.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java|11|18|61.11%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/BackendClause.java|6|6|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        @Override
+        public Void visitBackendClause(BackendClause backendClause, ConnectContext context) {
+            try {
+                for (String hostPort : backendClause.getHostPorts()) {
+                    Pair<String, Integer> pair = SystemInfoService.validateHostAndPort(hostPort);
+                    backendClause.getHostPortPairs().add(pair);
+                }
+                Preconditions.checkState(!backendClause.getHostPortPairs().isEmpty());
+            } catch (AnalysisException e) {
+                throw new RuntimeException(e);
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java:36]:
```suggestion
                throw new SemanticException(e);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        @Override
+        public Void visitFrontendClause(FrontendClause frontendClause, ConnectContext context) {
+            try {
+                Pair<String, Integer> pair = SystemInfoService.validateHostAndPort(frontendClause.getHostPort());
+                frontendClause.setHost(pair.first);
+                frontendClause.setPort(pair.second);
+                Preconditions.checkState(!Strings.isNullOrEmpty(frontendClause.getHost()));
+            } catch (AnalysisException e) {
+                throw new RuntimeException(e);
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java:49]:
```suggestion
                throw new SemanticException(e);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java [hunk 3]
```diff
@@ -170,6 +171,14 @@ public Void visitDropMaterializedViewStatement(DropMaterializedViewStmt statemen
         }
 
         @Override
+        public Void visitAlterSystemStmt(AlterSystemStmt statement, ConnectContext session) {
+            if (!GlobalStateMgr.getCurrentState().getAuth().checkGlobalPriv(ConnectContext.get(), PrivPredicate.OPERATOR)) {
+                ErrorReport.reportSemanticException(ErrorCode.ERR_SPECIFIC_ACCESS_DENIED_ERROR,
+                        "NODE");
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java:177]:
```suggestion
                       session.getQualifiedUser());
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 4]
```diff
@@ -42,17 +42,38 @@ statement
...
 tableRenameClause
     : RENAME identifier
     ;
 
+addBackendClause
+    : ADD FREE? BACKEND (TO string)? string (',' string)*
+    ;
+
+dropBackendClause
+    : DROP BACKEND string (',' string)*
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:66]:
```suggestion
    : DROP BACKEND string (',' string)* force?
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 5]
```diff
@@ -563,6 +570,54 @@ public ParseNode visitColumnNameWithComment(StarRocksParser.ColumnNameWithCommen
...
+
+    @Override
+    public ParseNode visitAddFrontendClause(StarRocksParser.AddFrontendClauseContext context) {
+        String cluster = ((StringLiteral) visit(context.string())).getStringValue();
+        if (context.FOLLOWER() != null) {
+            return new AddFollowerClause(cluster);
+        } else if (context.OBSERVER() != null) {
+            return new AddObserverClause(cluster);
+        } else {
+            return null;
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:606]: add Preconditions

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 6]
```diff
@@ -563,6 +570,54 @@ public ParseNode visitColumnNameWithComment(StarRocksParser.ColumnNameWithCommen
...
+
+    @Override
+    public ParseNode visitDropFrontendClause(StarRocksParser.DropFrontendClauseContext context) {
+        String cluster = ((StringLiteral) visit(context.string())).getStringValue();
+        if (context.FOLLOWER() != null) {
+            return new DropFollowerClause(cluster);
+        } else if (context.OBSERVER() != null) {
+            return new DropObserverClause(cluster);
+        } else {
+            return null;
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:618]: add Preconditions

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 [hunk 7]
```diff
@@ -185,6 +185,12 @@ WHERE: 'WHERE';
 WITH: 'WITH';
 YEAR: 'YEAR';
 FORCE: 'FORCE';
+SYSTEM: 'SYSTEM';
```
[the remark this answers is not in the document]
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4]: Solved.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 8]
```diff
@@ -542,25 +563,24 @@ number
...
+    | MAX | MIN | MINUTE | MONTH | MERGE | MATERIALIZED
     | NONE | NULLS
-    | OFFSET
+    | OFFSET | OBSERVER
     | PASSWORD | PRECEDING | PROPERTIES
     | ROLLUP | ROLLBACK | REPLICA
-    | SECOND | SESSION | SETS | START | SUM | STATUS
-    | TABLES | TABLET | TEMPORARY | TIMESTAMPADD | TIMESTAMPDIFF | THAN | TIME | TYPE
+    | SECOND | SESSION | SETS | START | SUM | STATUS | SYSTEM
+    | TABLES | TABLET | TEMPORARY | TIMESTAMPADD | TIMESTAMPDIFF | THAN | TIME | TYPE | TO
```
[the remark this answers is not in the document]
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: Removed the SYSTEM&TO

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 9]
```diff
@@ -42,17 +42,38 @@ statement
...
+    | addFrontendClause
+    | dropFrontendClause
     ;
 
 tableRenameClause
     : RENAME identifier
     ;
 
+addBackendClause
+    : ADD FREE? BACKEND (TO string)? string (',' string)*
```
[the remark this answers is not in the document]
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: Modify string and modify AstBuilder.java as the same time.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            return null;
+        }
+
+
+        public Void visitAlterSystemStmt(AlterSystemStmt statement, ConnectContext context) {
```
[the remark this answers is not in the document]
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java]: removed.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java (hunk 1 above)
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java]: modify to new SemanticException("backend host or port is wrong!")

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java (hunk 2 above)
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java]: modify to new SemanticException("frontend host or port is wrong!")

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java (hunk 3 above)
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java]: solved.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java (hunk 6 above)
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:622]: add.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java (hunk 5 above)
Comment (happut) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:609]: add.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 52 / 73 (71.23%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterSystemStmt.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|19|26|73.08%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java|18|23|78.26%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/FrontendClause.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/BackendClause.java|2|2|100.00%|

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 52 / 73 (71.23%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterSystemStmt.java|3|5|60.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|19|26|73.08%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterSystemStmtAnalyzer.java|18|23|78.26%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/FrontendClause.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/analysis/BackendClause.java|2|2|100.00%|

**Review by HangyuanLiu — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**